Update teleport build instructions (#15092)
gravitational/teleport: README.md

### Local Build

**Important:**

* The Go compiler is somewhat sensitive to the amount of memory: you will need
  **at least** 1GB of virtual memory to compile Teleport. A 512MB instance
  without swap will **not** work.
* This will build the latest version of Teleport, **regardless** of whether it
  is stable. If you want to build the latest stable release, run `git checkout`
  to the corresponding tag (for example, run `git checkout v8.0.0`) **before**
  performing a build.
* The following instructions are Mac-specific.

#### Dependencies

Ensure you have installed correct versions of dependencies:

> * `Go` version from [go.mod](https://github.com/gravitational/teleport/blob/0e9d621b464f2bd05a28e7db0ae2ef65349c460e/go.mod#L3)

Reviewer: Link to master? Picking a fixed commit is likely to yield outdated info over time.

```suggestion
* `Go` version from
  [go.mod](https://github.com/gravitational/teleport/blob/master/go.mod#L3)
```

Author: Yeah sorry I'll fix all links to point at master. Got these through "Open in Github" in IDE, which leads to my branch obviously.

>   ```shell

Reviewer: My 2c: I wouldn't prescribe specific instructions for a system. I think it's OK to say "you need Go version so-and-so, Rust and Cargo so-and-so" and let people do whatever makes sense for their setup. This is liable to get outdated and/or too opinionated, all the while missing a number of other systems anyway.

Author: See https://github.com/gravitational/teleport/pull/15092#issuecomment-1201867314

  # if we are not on the latest, you might need to
  # brew install go@<some_version>, i.e. 1.16
  #
  # check which version will be installed by running
  # brew info go
  
  $ brew install go
  ````
  
If you wish to build the Rust-powered features like Desktop Access:

> * `Rust` and `Cargo` version from [build.assets/Makefile](https://github.com/gravitational/teleport/blob/c3dde989cc9de6fa199f1a12287926218cc2e685/build.assets/Makefile#L21):

Reviewer: This one is tougher to link without fixing the commit, but maybe:

```suggestion
* `Rust` and `Cargo` version from
  [build.assets/Makefile](https://github.com/gravitational/teleport/blob/master/build.assets/Makefile#L21)
  (search for RUST_VERSION):
```

Author: I'll fix it

  
  ```shell
  $ brew install rustup
  $ rustup-init
  # accept defaults
  #
  # Once command finishes successfully, you might need to add
  # 
  # export PATH="$HOME/.cargo/bin:$PATH"
  # 
  # into ~/.zprofile and run
  # 
  # $ . ~/.zprofile
  # 
  # or open a new shell
  #
  $ rustup toolchain install <version from build.assets/Makefile>
  $ cd <teleport.git>
  $ rustup override set <version from build.assets/Makefile>
  $ rustc --version                                                                                                                                                                  ─╯
  rustc <version from build.assets/Makefile> (db9d1b20b 2022-01-20)
  ```
To build `tsh` version > `10.x` with `libfido` support:  

* Install `libfido2` and `openssl v 1.1.1` 
  ```shell

>   $ brew install libfido2
>   $ brew install openssl@1.1

Reviewer: Installing `openssl@1.1` explicitly is not necessary, but `export PKG_CONFIG_PATH="$(brew --prefix openssl@1.1)/lib/pkgconfig"` is.

As in other sections, I'd rather link folks to https://github.com/gravitational/teleport/blob/master/lib/auth/webauthncli/doc.go#L15 and https://github.com/Yubico/libfido2#installation, which have more complete instructions. We should also mention that it's necessary to pass `FIDO2=dynamic` to make for this to take effect.

```suggestion
  $ brew install libfido2
  $ export PKG_CONFIG_PATH="$(brew --prefix openssl@1.1)/lib/pkgconfig
  $ make build/tsh FIDO2=dynamic  # builds tsh with libfido2 support
```

Edit: I see you added some of this below, but some of the advice stands.

Author: Ah, I see, it pulls `openssl@1.1` as a dependency. Good point. And I am with you, will link to https://github.com/Yubico/libfido2#installation and https://github.com/gravitational/teleport/blob/master/lib/auth/webauthncli/doc.go#L15

> We should also mention that it's necessary to pass FIDO2=dynamic to make for this to take effect.

As in, explicitly in text, not just on the provided command line?

Reviewer: Just the make command line is good! I wrote this comment too soon.

  ```

> 

Reviewer: Maybe add:

```suggestion
`tsh` binaries with Touch ID support are only functional using binaries signed
with Teleport's Apple Developer ID and notarized by Apple. If you are a Teleport
maintainer, ask the team for access. The make switch for Touch ID is
`TOUCHID=yes`.

```

if you think it's worth a nod.

Author: Noice, yes for sure. Thanks

#### Perform a build

Get the source

```shell
$ git clone https://github.com/gravitational/teleport.git
$ cd teleport
```

To perform a build 

```shell
$ make full
```

To build `tsh` with `libfido`

* On M1 Mac

  ```shell

>   PKG_CONFIG_PATH=/opt/homebrew/opt/openssl@1.1/lib/pkgconfig make build/tsh FIDO2=dynamic

Reviewer: Pasted a universal snippet above.

Author: Nice, thanks!

  ```

* On Intel Mac

  ```shell
  PKG_CONFIG_PATH=/usr/local/opt/openssl@1.1/lib/pkgconfig make build/tsh FIDO2=dynamic
  ```

#### Build output and running locally

If the build succeeds, the installer will place the binaries in the `build` directory.

> Before starting, create default data directories:

Reviewer: I'd much prefer changing the data_dir and not relying on `/var/lib/teleport` for dev builds, but that's me.

Author: This is part of original docs, I just moved the text around. Not sure if we want to make a chance in the scope of this PR.

Reviewer: That's OK, no need to change anything here.

### Local Build

**Important:**

* The Go compiler is somewhat sensitive to the amount of memory: you will need
  **at least** 1GB of virtual memory to compile Teleport. A 512MB instance
  without swap will **not** work.
* This will build the latest version of Teleport, **regardless** of whether it
  is stable. If you want to build the latest stable release, run `git checkout`

>   to the corresponding tag (for example, run `git checkout v8.0.0`) **before**

Reviewer 2: Might want to add `git submodule --update --recursive` as well to make sure web assets are correct.

Author: This is entirely specific to Teleport employees though, right? Do we want to include this into public repo README?

Reviewer 3: The webassets submodule is public, so I would say yes.

Author: But not the `e` submodule, right?

Reviewer 3: Right, but `e` is not required to do an OSS teleport build.

Author: Sure, however `git submodule --update --recursive` will fail for anyone but Teleport Employees?

Author: Anyways, I added a note about it.

  performing a build.
* The following instructions are Mac-specific.

#### Dependencies

Ensure you have installed correct versions of dependencies:

* `Go` version from
  [go.mod](https://github.com/gravitational/teleport/blob/master/go.mod#L3)
  
  ```shell
  # if we are not on the latest, you might need to
  # brew install go@<some_version>, i.e. 1.16
  #
  # check which version will be installed by running
  # brew info go
  
  $ brew install go

>   ````

Reviewer 2: Homebrew is not heavily used by Go users. Maybe instead link to the Go Download and Install instructions? https://go.dev/doc/install

Author: Sure, I can add that. Any particular reason homebrew is not used by Go users? And what is used in favor - just downloading Go and managing versions via `go install ...` ?

Reviewer 3: I have used Go from homebrew for years. 🤷‍♂️

Reviewer: Me too, that's usually I what I tell people to do.

If we keep it generic here and move macOS/brew-specific instructions to a separate file, I think this is kind of a moot point. We can just say there "here is a macOS/Homebrew-based setup" and that's it.

Author: I moved the instructions into a separate file in https://github.com/gravitational/teleport/pull/15092/commits/66ed09fdf93c03d922bc62cc061752b20a36e098 and provided setup for both homebrew and go.dev download.

### Local Build

**Important:**

* The Go compiler is somewhat sensitive to the amount of memory: you will need
  **at least** 1GB of virtual memory to compile Teleport. A 512MB instance
  without swap will **not** work.
* This will build the latest version of Teleport, **regardless** of whether it
  is stable. If you want to build the latest stable release, run `git checkout`
  to the corresponding tag (for example, run `git checkout v8.0.0`) **before**
  performing a build.
* `tsh` binaries with Touch ID support are only functional using binaries signed
  with Teleport's Apple Developer ID and notarized by Apple. If you are a Teleport
  maintainer, ask the team for access.

#### Dependencies

Ensure you have installed correct versions of necessary dependencies:
* `Go` version from
  [go.mod](https://github.com/gravitational/teleport/blob/master/go.mod#L3)

> * If you wish to build the Rust-powered features like Desktop Access, `Rust` and `Cargo` version from
>   [build.assets/Makefile](https://github.com/gravitational/teleport/blob/master/build.assets/Makefile#L21)
>   (search for `RUST_VERSION`)

Reviewer: ```suggestion
* If you wish to build the Rust-powered features like Desktop Access, see the
  `Rust` and `Cargo` version in
  [build.assets/Makefile](https://github.com/gravitational/teleport/blob/master/build.assets/Makefile#L21)
  (search for `RUST_VERSION`)
```

* For `tsh` version > `10.x` with FIDO support, you will need `libfido` and `openssl 1.1` installed locally

For an example of Dev Environment setup on a Mac, see [these instructions](README_local_MacOS.md). 

#### Perform a build

Get the source

```shell
$ git clone https://github.com/gravitational/teleport.git
$ cd teleport
```

To perform a build 

```shell
$ make full
```

> To build `tsh` with Apple TouchID support enabled:

Reviewer: My 2c: I would ditch this or just mention the build tag in the "Important" section above. I'm afraid that, if we leave this here, folks will try to build with Touch ID support and not see/ignore the warning.

Author: Yeah, you are right. I'll see if we can consolidate these instructions and the note about TouchID/Apple developer thing.

Author: Moved the section around in https://github.com/gravitational/teleport/pull/15092/commits/a33951a6a363735adf3ef7a6058f30344c6122a3
